### ISSUE TITLE
Release the gil in linalg module

### ISSUE THREAD

**Issue by jcrist** [numpy/numpy#7179]
Currently, calls to routines in `np.linalg` don't release the interpreter lock, but (as far as I can tell) should be able to. This has caused non-optimal behavior for [dask.array.linalg](https://github.com/blaze/dask/blob/master/dask/array/linalg.py) when trying to use these functions in parallel using threads.

**Comment by jakirkham** [comment]
Could you explain a bit more what you mean? Many of these functions call BLAS or LAPACK routines that have the GIL release before they are called. Maybe an example of a function you are using and the behavior you are seeing would be constructive here.

**Comment by seberg** [comment]
I am not sure they do, they all go through the Generalized ufunc machinery, it *should* release the GIL, but frankly, a quick search did not show a release in the gufunc machinery for me, and I do not think linalg will do this itself (or probably should have to).

I did not test it, but I assume there is some check for API being necessary + releasing it if not path missing in the generalized ufunc outer loop. It would probably be very similar to the code for normal ufuncs.

**Comment by jcrist** [comment]
They definitely don't.

```python
In [1]: import numpy as np

In [2]: from multiprocessing.pool import ThreadPool

In [3]: p = ThreadPool()

In [4]: np.__version__
Out[4]: '1.10.4'

In [5]: a = np.random.random((2000, 2000))

In [6]: a = a * a.T

In [7]: %time _ = np.linalg.qr(a)
CPU times: user 849 ms, sys: 37.4 ms, total: 886 ms
Wall time: 884 ms

In [8]: %time _ = p.map(np.linalg.qr, [a, a])
CPU times: user 1.72 s, sys: 83.9 ms, total: 1.81 s
Wall time: 1.61 s
```

**Comment by jakirkham** [comment]
Thanks for the example.

**Comment by seberg** [comment]
I think it is missing likely, though performance wise, I am not sure you will have any (or even negative) gain over using a threaded BLAS library (which you might already be doing).

**Comment by jcrist** [comment]
In the example code I posted above I'd turned the threading off for my BLAS. I don't doubt that numpy + dask will be slower than a threaded BLAS, the benefit of using dask for parallelizing these will mostly be the ability to work out-of-core. I'd like to be able to exploit parallelism here when possible though, which is currently prevented by the gil (for an example of the effect this has, see the plot/discussion on QR here: http://dask.pydata.org/en/latest/diagnostics.html#example).

**Comment by seberg** [comment]
Yes, we should definitely get it fixed, it might even be a regression, though a long standing one then.

**Comment by jcrist** [comment]
If this is a simple enough fix, I'd be happy to tackle it. I'm unfamiliar with the numpy codebase though, and wouldn't know where to start. This also isn't super pressing, just something I noticed while profiling :).

**Comment by jakirkham** [comment]
If you needed something in the interim, I believe `dask` has some way to harness `ipyparallel` (was IPython Cluster) to farm the work out to different processes.

**Comment by seberg** [comment]
You can try gh-7198 if you like. Ironically, QR is *not* part of it, because unlike most of the functions, it is not a generalized ufunc....

**Comment by charris** [comment]
@seberg Can this be closed?

**Comment by seberg** [comment]
I think QR still is not a gufunc, and it would probably be nice if it was, so renaming the issue into that.
